File: packages/sealedrun-py/src/sealedrun/hashing.py
"""Hash algorithms of SPEC 4.1 and the base64url and hex encodings of SPEC 3."""

from __future__ import annotations

import base64
import hashlib
from typing import Any, Literal

from sealedrun.canonical import canonicalize

HashAlg = Literal["sha-256", "sha-384"]
HASH_ALGS: dict[str, tuple[str, int]] = {"sha-256": ("sha256", 32), "sha-384": ("sha384", 48)}
DEFAULT_HASH_ALG: HashAlg = "sha-256"


def digest(alg: str, data: bytes) -> bytes:
    """Hash `data` with a registered `hash_alg`; an unregistered name raises KeyError."""
    name, _ = HASH_ALGS[alg]
    return hashlib.new(name, data).digest()


def digest_size(alg: str) -> int:
    """Return the digest length of `alg` in bytes."""
    return HASH_ALGS[alg][1]


def zero_hash(alg: str) -> str:
    """Return the all-zero hex hash that the first record of a run uses as `prev_hash`."""
    return "0" * (digest_size(alg) * 2)


def object_hash(alg: str, protected: dict[str, Any]) -> str:
    """Hash the JCS form of the protected part of an object and return lowercase hex."""
    return digest(alg, canonicalize(protected)).hex()


def b64url_encode(data: bytes) -> str:
    """Encode bytes as base64url without padding (RFC 4648 section 5)."""
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode("ascii")


def b64url_decode(text: str) -> bytes:
    """Decode base64url text that carries no padding."""
    padding = "=" * (-len(text) % 4)
    return base64.urlsafe_b64decode(text + padding)


def payload_digest(alg: str, body: bytes) -> str:
    """Return the base64url digest of a payload body, as stored in a payload reference."""
    return b64url_encode(digest(alg, body))

File: packages/sealedrun-ts/src/hashing.ts
import { sha256, sha384 } from "@noble/hashes/sha2.js";

import { canonicalBytes, type Json } from "./canonical.js";
import { b64urlEncode, hexEncode } from "./encoding.js";

/** Hash algorithms registered in SPEC 4.1. */
export type HashAlg = "sha-256" | "sha-384";

const SIZES: Record<HashAlg, number> = { "sha-256": 32, "sha-384": 48 };

/** Narrows an untrusted `hash_alg` string to a supported algorithm. */
export function isHashAlg(value: string): value is HashAlg {
  return Object.hasOwn(SIZES, value);
}

/** Raw digest of the data. */
export function digest(alg: HashAlg, data: Uint8Array): Uint8Array {
  return alg === "sha-256" ? sha256(data) : sha384(data);
}

/** Digest length in bytes. */
export function digestSize(alg: HashAlg): number {
  return SIZES[alg];
}

/** The all-zero hex string used as `prev_hash` of the record with seq 0 (SPEC 5.3). */
export function zeroHash(alg: HashAlg): string {
  return "0".repeat(SIZES[alg] * 2);
}

/**
 * Hex digest of the canonical form of the protected part of an object (SPEC 5.3).
 *
 * @throws TypeError if the object contains a non-finite number.
 */
export function objectHash(alg: HashAlg, protectedObject: Json): string {
  return hexEncode(digest(alg, canonicalBytes(protectedObject)));
}

/** Base64url digest of a payload body or bundle file, as stored in `payload` and `files`. */
export function payloadDigest(alg: HashAlg, body: Uint8Array): string {
  return b64urlEncode(digest(alg, body));
}
